Initial insert of jte integration

casid opened the pull request:
Hi @tipsy, this is a first draft of a jte integration for Javalin!

I went with solution 1, and added a "jte" key to the model map, which is a bit hacky.

Changing the model parameter to `Any`, would be nicer for jte, but require a change to the interface. Not sure if you would want to do that.

Maybe there's an even better solution I can't think of right now?

tipsy commented:
Thanks, that was fast. I'll have a look in the weekend !

tipsy commented:
Great work @casid. I agree that `ctx.render("test.jte", mapOf("jte" to JteTestPage("hello", "world")))` looks a bit hacky, and I don't think we should recommend that approach. We could change the interface, but that would make the experience worse for all the other template engines. What do you think about doing somehting like:

```kotlin
ctx.render("test.jte", Jte.page(JteTestPage("hello", "world")))
```

Where `Jte.page` is a static method that returns a map `{"jte": page}`?

Alternatively we could create `ctx.render(Page)` which does the same thing:

```kotlin
fun render(filePath: String, page: JtePage = Jte.emptyPage): Context {
    return html(JavalinRenderer.renderBasedOnExtension(filePath, mapOf("jte", page), this))
} 
```

In which case you would be able to do `ctx.render("test.jte", JteTestPage("hello", "world"))`. That is a tighter coupling that I would ideally want though 🤔

What do you think? Any other ideas?

casid commented:
Thanks for looking into it @tipsy!

I agree, having jte in the context feels bad and tightly coupled! I thought about it the last days and maybe there are some other ways. 

### Idea 1
What do you think about adding a second method to `FileRenderer`, that takes Any as parameter? `FileRenderer` wouldn't be a `FunctionalInterface` anymore, but I haven't seen it used like that anywhere in the project.

```kotlin
public interface FileRenderer {
    String render(@NotNull String filePath, @NotNull Map<String, Object> model, @NotNull Context context) throws Exception;
    String render(@NotNull String filePath, @NotNull Object model, @NotNull Context context) throws Exception;
}
```

And have an overload in `Context` like this:

```kotlin
@JvmOverloads
fun render(filePath: String, model: Any): Context {
     return html(JavalinRenderer.renderBasedOnExtension(filePath, model, this))
}
```

There would be some engines that could handle this implementation:

- Freemarker (I used it before I wrote jte and it can use objects instead of maps)
- Mustache (at least the interface looks like it could work with an Object)

Problems I see with this approach is that it might be confusing for the user what method to use. Also the implementations not supporting it would need to throw an exception (and the other way in jte's case). On the other hand, wrapping models around some jteMap doesn't allow to switch to another engine either. With this implementation you could at least switch between Freemarker and jte ;-)

### Idea 2

Take an approach like `JavalinVue` and create a class `JavalinJte` apart from the other template engines. Because right now all template engines are simple wrappers around the particular engines and don't do much beyond that point. And I think Javalin could do a lot more if you want it to.

At work we are currently evaluating to replace JSP with jte. Our web framework there has a lot of custom JSP tags that help with form validations and submissions. For that I've added the ability in jte 0.5 to intercept HTML elements at compile time with hooks for adding additional HTML at runtime. This means the frontend guy can write regular <form>, <input> tags without having to deal with any framework specific tags. For instance:

```html
<form>
    <input type="text" name="email">
</form>
```

With the right config, jte would render the output like this, based on the controller's current model value of `email`:

```html
<form>
    <input type="text" name="email" value="test@example.com">
</form>
```

Our project at work is closed source, but you can have a look at <a href="https://github.com/casid/jte/blob/master/src/test/java/org/jusecase/jte/TemplateEngine_HtmlTagSupportTest.java">TemplateEngine_HtmlTagSupportTest</a> to get an idea what's possible. It fits really well with jte's philosophy to utilize well known language features over new syntax to learn :-)

Maybe that'd be a pretty nifty feature to bring to Javalin? I know Javalin's philosiphy not to need to extend anything, but I thought if you look a this stuff, you might instantly see a use case for it ;-)

What do you think?

tipsy commented:
> Problems I see with this approach is that it might be confusing for the user what method to use

Yup, that's something I would really like to avoid!

Going with the `JavalinVue` approach we'd have:

```kotlin
app.get("/hello", JtePage("/path/to/file.jte", page)) // page being optional - unlikely to be used, maybe for static pages?
```

and 

```kotlin
app.get("/hello") { ctx ->
    JtePage("/path/to/file.jte", page).handle(ctx)
}
```

The tradeoff here is that we lose dicoverability, but that's not really a big concern IMO, as the current implementation relies on file-extensions anyway. The syntax isn't the best looking, but it's probably better than changing the existing codebase.

> Maybe that'd be a pretty nifty feature to bring to Javalin? I know Javalin's philosiphy not to need to extend anything, but I thought if you look a this stuff, you might instantly see a use case for it ;-)
> What do you think?

I haven't used server side templates in many years, so I'm unfortunately a bit lacking in imagination, but I'm okay with using inheritance in this case. I don't really see why it should make a difference thought, maybe you could create an example?

casid commented:
Thanks for the fast reply! I'll try to write a small example during the next week :-)

casid commented:
@tipsy I thought about a server side form validation example, but I must admit that this would be a pretty big project to add to Javalin. And I'm sure most devs do that stuff with ajax calls nowadays anyways.

But! I've updated jte to be compatible with Javalin's `FileRenderer` interface. As of jte 0.6.0 all root templates are allowed to have more than one parameter. So there's now an additional method that takes a `Map<String, Object>` that is then mapped to the parameters defined in the template. It comes with close to zero overhead as well, since the compiler now generates an additional mapping function for each template (even supporting default values for params!) like this:

```java
public static void renderMap(org.jusecase.jte.TemplateOutput output, org.jusecase.jte.support.HtmlTagSupport htmlTagSupport, java.util.Map<String, Object> params) {
	boolean firstParam = (boolean)params.getOrDefault("firstParam", false);
	Boolean secondParam = (Boolean)params.getOrDefault("secondParam", true);
	String thirdParam = (String)params.get("thirdParam");
	render(output, htmlTagSupport, firstParam, secondParam, thirdParam);
}
```

That'd be for a jte template like this:

```xml
@param boolean firstParam = false
@param Boolean secondParam = true
@param String thirdParam

template stuff...
```

I did some benchmarks and there is no measurable performance impact compared to the single param usage.

I've updated the PR with jte 0.6.0 and adjusted the tests and `JavalinJte`.

Let me know what you think and if you want me to do any other changes before merging!

casid commented:
Heya @tipsy, did you have a chance to look into the Javalin jte integration so far? Let me know if there is anything I should change.

Also, I'm heavily working on the next jte release. Starting with 0.8, jte understands HTML templates at compile time, providing context-sensitive, pre-compiled and super fast output escaping. Still having 90k templates/sec on the mbosecke/template-benchmark with html mode and full output escaping 👍 

0.8 requires the user to choose in `TemplateEngine.create()` what content type all templates managed by that engine have. As soon as 0.8 is ready I would suggest to adjust the Javalin default jte engine to be `ContentType.Html`, since that's probably what 99% of all users will do with it.

tipsy commented:
@casid currently on vacation. I'll have a look in the weekend :)

casid commented:
No rush, sorry for disturbing! Enjoy your vacation :-)

tipsy commented:
Looks good @casid! I like the support for multiple params. Do you want to wait for 0.8, or should I merge it right away?

casid commented:
Thanks @tipsy! You can merge right away. 0.8 Is probably gonna take me another month or so to finish.

Would it be alright if I do another PR with 0.8, once everything is smooth and running with HTML escaping?

tipsy commented:
> Would it be alright if I do another PR with 0.8, once everything is smooth and running with HTML escaping?

That would be great. It would also be nice to have an official server side templating tutorial (showcasing your IntelliJ plugin). Would you be interested in writing something like that for https://javalin.io/tutorials ?
